extensions/redis-cache/src/main/java/io/druid/client/cache/RedisCache.java

```diff
@@ -0,0 +1,288 @@
+      else {
+        missCount.incrementAndGet();
+      }
+    }
+    catch (JedisConnectionException ex) {
+      if (isTimeout(ex)) {
+        timeoutCount.incrementAndGet();
+      }
+
+      throw Throwables.propagate(ex);
+    }
+    catch (JedisException | LZ4Exception ex) {
+      errorCount.incrementAndGet();
+      log.warn(ex, "Error getting cache item");
+      return null;
+    }
+
+    return result;
+  }
+
+  @Override
+  public void put(NamedKey key, byte[] value)
+  {
+    final byte[] hash = computeKeyHash(prefix, key);
+    final byte[] serialized;
+
+    try {
+      serialized = serialize(key, value);
+    }
+    catch (NullPointerException ex) {
+      throw Throwables.propagate(ex);
+    }
+
+    try (final Jedis client = pool.getResource()) {
+      client.psetex(hash, expiration, serialized);
+    }
+    catch (JedisConnectionException ex) {
+      Throwables.propagate(ex);
+    }
+    catch (JedisException ex) {
+      errorCount.incrementAndGet();
+      log.warn(ex, "Error putting cache item");
+    }
+  }
+
+  @Override
+  public Map<NamedKey, byte[]> getBulk(Iterable<NamedKey> keys)
+  {
+    final Map<byte[], NamedKey> keyLookup = Maps.uniqueIndex(
+      keys,
+      new Function<NamedKey, byte[]>()
+      {
+        @Override
+        public byte[] apply(
+          @Nullable NamedKey input
+        )
+        {
+          return computeKeyHash(prefix, input);
+        }
+      }
+    );
+    final Map<NamedKey, byte[]> results = Maps.newHashMap();
+
+    // Sometimes broker passes empty keys list to getBulk()
+    if (keyLookup.size() == 0) {
+      return results;
+    }
+
+    final List<byte[]> keysList = new ArrayList<>(keyLookup.keySet());
+    final List<byte[]> valuesList;
+
+    try (final Jedis client = pool.getResource()) {
+      valuesList = client.mget(keysList.toArray(new byte[keysList.size()][]));
+    }
+    catch (JedisConnectionException ex) {
+      if (isTimeout(ex)) {
+        timeoutCount.incrementAndGet();
+      }
+
+      throw Throwables.propagate(ex);
+    }
+    catch (JedisException ex) {
+      errorCount.incrementAndGet();
+      log.warn(ex, "Unable to get cache items");
+      return results;
+    }
+
+    final int length = valuesList.size();
+    int i = 0;
+
+    while (i < length) {
+      final byte[] value = valuesList.get(i);
+      if (value != null) {
+        final NamedKey key = keyLookup.get(keysList.get(i));
+        try {
+          final byte[] deserialized = deserialize(key, value);
+          hitCount.incrementAndGet();
+          results.put(key, deserialized);
+        }
+        catch (LZ4Exception ex) {
+          errorCount.incrementAndGet();
+          log.warn(ex, "Error decompressing cache item");
+        }
+      }
+      else {
+        missCount.incrementAndGet();
+      }
+
+      i++;
+    }
+
+    return results;
+  }
+
+  @Override
+  public void close(String namespace)
+  {
+    final byte[] pattern = redisKey(prefix, hash(namespace), "*");
```

Comment:
From performance point of view maybe it would be better to have a list per namespace and whenever we do `Cache.put()` put another hash into this list. So once we issue `Cache.close()` we do LRANGE key 0 -1 and then DEL. That might be much faster.

If the issue is in thread blocking, then I must say Jedis does not support async at this time though development of async client is in progress